miner/src/service_transaction_checker.rs

```diff
@@ -16,37 +16,111 @@
 
+/// Certified addresses cache
+#[derive(Default)]
+struct CertifiedAddressesCache {
+	cache: RwLock<BTreeMap<Address, bool>>,
+	last_block: u64
+}
+
 /// Service transactions checker.
 #[derive(Default, Clone)]
-pub struct ServiceTransactionChecker;
+pub struct ServiceTransactionChecker {
+	certified_addresses_cache: CertifiedAddressesCache
+}
 
 impl ServiceTransactionChecker {
-	/// Checks if given address is whitelisted to send service transactions.
-	pub fn check<C: CallContract + RegistryInfo>(&self, client: &C, tx: &SignedTransaction) -> Result<bool, String> {
+	/// Checks if given address in tx is whitelisted to send service transactions.
+	pub fn check<C: CallContract + RegistryInfo + ChainInfo>(&self, client: &C, tx: &SignedTransaction) -> Result<bool, String> {
 		let sender = tx.sender();
-		let hash = tx.hash();
-
 		// Skip checking the contract if the transaction does not have zero gas price
 		if !tx.gas_price.is_zero() {
 			return Ok(false)
 		}
 
-		let address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
+		self.check_address(client, sender)
+	}
+
+	/// Checks if given address is whitelisted to send service transactions.
+	pub fn check_address<C: CallContract + RegistryInfo + ChainInfo>(&self, client: &C, sender: Address) -> Result<bool, String> {
+		let cache = self.certified_addresses_cache.cache.try_read();
+		if (client.chain_info().best_block_number as u64 == self.certified_addresses_cache.last_block) && cache.is_some() {
+			if let Some(allowed) = cache.unwrap().get(&sender) {
+				return Ok(*allowed);
+			}
+		}
+		let contract_address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
 			.ok_or_else(|| "contract is not configured")?;
+		trace!(target: "txqueue", "Checking service transaction checker contract from {}", sender);
+		let (data, decoder) = service_transaction::functions::certified::call(sender);
+		let value = client.call_contract(BlockId::Latest, contract_address, data)?;
+		decoder.decode(&value).and_then(|allowed| {
+			let cache = self.certified_addresses_cache.cache.try_write();
+			if cache.is_some() {
+				cache.unwrap().insert(sender, allowed);
+			};
+			Ok(allowed)
+		}).map_err(|e| e.to_string())
+	}
 
-		trace!(target: "txqueue", "[{:?}] Checking service transaction checker contract from {}", hash, sender);
+	/// Refresh certified addresses cache
+	pub fn refresh_cache<C: CallContract + RegistryInfo + ChainInfo>(&mut self, client: &C) -> Result<bool, String> {
+		trace!(target: "txqueue", "Refreshing certified addresses cache");
+		let contract_address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest);
+		if contract_address.is_none() {
+			return Ok(false)
+		}
+		let mut updated_addresses: BTreeMap<Address, bool> = BTreeMap::default();
+		let cache = self.certified_addresses_cache.cache.try_read();
+		if cache.is_some() {
```

Comment:
Using `try_read()` here means that we won't update the cache if we are unlucky. I'd rather go for more predictable behavior.

Reply:
I used `try_read()` and `try_write()` because it was not clear to me how many threads will be using the cache, as far as I can see now - there will be two of them: one for client and one for miner, so I'm still not sure how to guarantee that cache won't be accessed while another thread will be writing into it.

I can think about implementing some waiting when the cache is busy, but still I doubt whether it's a proper solution.

Reply:
Sorry, I think I misunderstood a bit how the RwLock works. I have done some tests now and I saw that even if the cache is used by another thread writing into it, the `read` or `write` won't fail and they will be just waiting for the lock to release, so I believe it's safe to use `read` and `write` instead of `try_read` and `try_write`. Correct me please if I'm wrong.

Reply:
@VladLupashevskyi that's correct `read()` will block until it can acquire the lock, `try_read()` is a non-blocking version of it, but may fail in case cache is being currently updated (i.e. someone holds `.write()` lock).
While I think `try_read` is fine when we are checking the cache (if we are unlucky it means that the cache is being updated anyway - so probably it contains obsolete data), updating should block.

Regarding threads - we actually run inside a threadpool, so even though we have two structs using this checker (`Miner`/`Client`) they might be invoked on multiple threads in the pool in the same time.